lib/steep/type_construction.rb

```diff
@@ -2401,7 +2401,18 @@ def synthesize(node, hint: nil, condition: false)
             if value_node
               type, constr = synthesize(value_node, hint: hint)
 
-              if hint.is_a?(AST::Types::Proc) && value_node.type == :sym
+              if hint.is_a?(AST::Types::Proc) && value_node.type == :send && value_node.children[1] == :method && AST::Builtin::Method.instance_type?(type)
+                method_name = value_node.children[2].children[0] #: Symbol
+                if method = calculate_interface(AST::Types::Self.instance, private: true)&.methods&.[](method_name)
+                  if overload = method.method_types.find {|method_type| method_type.type.params.nil? || method_type.type.params.optional? }
+                    type = AST::Types::Proc.new(
+                      type: method.method_types.first&.type || raise,
```

Comment:
Is it better to use `overload` here?
```suggestion
                      type: overload,
```